lib/charms/mlops_libs/v0/k8s_service_info.py

```diff
@@ -0,0 +1,297 @@
+
+### Fetching the library with charmcraft
+
+Using charmcraft you can:
+```shell
+charmcraft fetch-lib charms.mlops_libs.v0.k8s_service_info
+```
+
+## Using the library as requirer
+
+### Add relation to metadata.yaml
+```yaml
+requires:
+  k8s-svc-info:
+    interface: k8s-service
+    limit: 1
+```
+
+### Instantiate the KubernetesServiceInfoRequirer class in charm.py
+
+```python
+from ops.charm import CharmBase
+from charms.mlops_libs.v0.kubernetes_service_info import KubernetesServiceInfoRequirer, KubernetesServiceInfoRelationError
+
+class RequirerCharm(CharmBase):
+    def __init__(self, *args):
+        self._k8s_svc_info_requirer = KubernetesServiceInfoRequirer(self)
+        self.framework.observe(self.on.some_event_emitted, self.some_event_function)
+
+    def some_event_function():
+        # use the getter function wherever the info is needed
+        try:
+            k8s_svc_info_data = self._k8s_svc_info_requirer.get_k8s_svc_info()
+        except KubernetesServiceInfoRelationError as error:
+            "your error handler goes here"
+```
+
+## Using the library as provider
+
+### Add relation to metadata.yaml
+```yaml
+provides:
+  k8s-svc-info:
+    interface: k8s-service
+```
+
+### Instantiate the KubernetesServiceInfoProvider class in charm.py
+
+```python
+from ops.charm import CharmBase
+from charms.mlops_libs.v0.kubernetes_service_info import KubernetesServiceInfoProvider, KubernetesServiceInfoRelationError
+
+class ProviderCharm(CharmBase):
+    def __init__(self, *args, **kwargs):
+        ...
+        self._k8s_svc_info_provider = KubernetesServiceInfoProvider(self)
+        self.observe(self.on.some_event, self._some_event_handler)
+    def _some_event_handler(self, ...):
+        # This will update the relation data bag with the Service name and port
+        try:
+            self._k8s_svc_info_provider.send_k8s_svc_info(svc_name, svc_port)
+        except KubernetesServiceInfoRelationError as error:
+            "your error handler goes here"
+```
+
+## Relation data
+
+The data shared by this library is:
+* svc_name: the name of the Kubernetes Service
+  as it appears in the resource metadata, e.g. "metadata-grpc-service".
+* svc_port: the port of the Kubernetes Service
+"""
+
+import logging
+from typing import List, Optional, Union
+
+from ops.charm import CharmBase
+from ops.framework import BoundEvent, Object
+from ops.model import Relation
+
+# The unique Charmhub library identifier, never change it
+LIBID = "f5c3f6cc023e40468d6f9a871e8afcd0"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# Default relation and interface names. If changed, consistency must be kept
+# across the provider and requirer.
+DEFAULT_RELATION_NAME = "k8s-svc-info"
+DEFAULT_INTERFACE_NAME = "k8s-service"
+REQUIRED_ATTRIBUTES = ["svc_name", "svc_port"]
```

Comment:
Since the object is called `KubernetsServiceInfo`, do we need to prefix these with `svc`?  I think `name` and `port` would be sufficient.

Reply:
Okay, will update it.

Reply:
yeah I see where you're coming from.  to me the clarity comes from naming the instance in some sane way.  like how this still makes sense:

```python
service_info = KubernetsServiceInfo()
service_info.name  # service name
```

but of course this would be terrible :) :

```python
charm = KubernetsServiceInfo()
charm.name  # service name, not charm name
```

But it is up to you

Reply:
Oh that is going to happen. Because we have the KubernetsServiceInfoObject, you'll get exactly that. So in your charm it will look like:

```
k8s_svc_info_requirer = KubernetesServiceInfoRequirer(...)
svc_name = k8s_svc_info_requirer.get_data().name
svc_port = k8s_svc_info_requirer.get_data().port
```

Reply:
Oh yes I see what you mean - these `REQUIRED_ATTRIBUTES` names are only used inside the databag anyhow.  Ok, from the user experience standpoint I no longer object as this is **almost** a hidden detail (almost, because a user unpacking relation data with `show-unit`, etc, will still see it.  but that's small).

I'd still prefer the content inside the relation data be named the same as the data model for different reasons though.  The pydantic data model for the lib code says `KubernetesServiceInfo` has attributes of `name` and `port`, but when you add a pydantic model for the relation in [charm-relation-interface repo](https://github.com/canonical/charm-relation-interfaces) the pydantic model will be different and instead have `svc_` suffixes.  That feels funny.

As is, the lib will work, so its not a hard blocker on this.  But having two data models for something that is created fresh feels unnecessary.

Reply:
Oh I hadn't noticed in the previous message, but things like `send_data()` still have the suffix so the user still needs to care.

This implementation feels asymmetric.  I'd expect something like:
```python
k8s_service_info_provider_wrapper.send_data(name='foo', port='bar')
service_info = k8s_service_info_requirer_wrapper.get_data()
print(service_info.name)  # uses the same nomenclature as we sent
print(service_info.port)
```

where the attribute going in has the same name as coming out, but instead we have:

```python
k8s_service_info_provider_wrapper.send_data(svc_name='foo', svc_port='bar')  # <--- `svc_` prefix
service_info = k8s_service_info_requirer_wrapper.get_data()
print(service_info.name)  # no prefix
print(service_info.port)
```

As a user of this lib, I don't want to remember that one side of the relation has a prefix while the other doesn't.

Reply:
I'll change the attributes to `name` and `port`.